### PR TITLE
modernize porkbun-ddns

### LINKED ISSUE #2: JSON Decode Error
I set the script up to run on reboot to update a bunch of domains. It worked for a few of them, but most failed.
This is a snippet from my CRON log:
```
Subject: Cron <k****u@the-server> /bin/python3 /External/porkbun-dynamic-dns-python/porkbun-ddns.py /External/porkbun-dynamic-dns-python/config.json s*******k.cloud -i ${myip}
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit
X-Cron-Env: <SHELL=/bin/sh>
X-Cron-Env: <HOME=/home/k****u>
X-Cron-Env: <PATH=/usr/bin:/bin>
X-Cron-Env: <LOGNAME=k****u>
Message-Id: <20210621052204.A7A2CEC0D3@the-server.T****h>
Date: Mon, 21 Jun 2021 00:22:04 -0500 (CDT)

Deleting existing A Record
Creating record: sudoz***k.cloud with answer of 67.***.***.217
Traceback (most recent call last):
  File "/External/porkbun-dynamic-dns-python/porkbun-ddns.py", line 50, in <module>
    print(createRecord()["status"])
  File "/External/porkbun-dynamic-dns-python/porkbun-ddns.py", line 28, in createRecord
    create = json.loads(requests.post(apiConfig["endpoint"] + '/dns/create/'+ rootDomain, data = json.dumps(createObj)).text)
  File "/usr/lib/python3.8/json/__init__.py", line 357, in loads
    return _default_decoder.decode(s)
  File "/usr/lib/python3.8/json/decoder.py", line 337, in decode
    obj, end = self.raw_decode(s, idx=_w(s, 0).end())
  File "/usr/lib/python3.8/json/decoder.py", line 355, in raw_decode
    raise JSONDecodeError("Expecting value", s, err.value) from None
json.decoder.JSONDecodeError: Expecting value: line 1 column 1 (char 0)

From k****u@the-server.T****h  Mon Jun 21 00:22:04 2021
Return-Path: <k****u@the-server.T****h>
X-Original-To: k****u
Delivered-To: k****u@the-server.T****h
Received: by the-server.T****h (Postfix, from userid 1000)
        id DBA18EC0AE; Mon, 21 Jun 2021 00:22:03 -0500 (CDT)
From: root@the-server.T****h (Cron Daemon)
To: k****u@the-server.T****h
Subject: Cron <k****u@the-server> /bin/python3 /External/porkbun-dynamic-dns-python/porkbun-ddns.py /External/porkbun-dynamic-dns-python/config.json z***k.download -i ${myip}
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit
X-Cron-Env: <SHELL=/bin/sh>
X-Cron-Env: <HOME=/home/k****u>
X-Cron-Env: <PATH=/usr/bin:/bin>
X-Cron-Env: <LOGNAME=k****u>
Message-Id: <20210621052203.DBA18EC0AE@the-server.T****h>
Date: Mon, 21 Jun 2021 00:22:03 -0500 (CDT)

Traceback (most recent call last):
  File "/External/porkbun-dynamic-dns-python/porkbun-ddns.py", line 47, in <module>
    myIP=getMyIP() #otherwise use the detected exterior IP address
  File "/External/porkbun-dynamic-dns-python/porkbun-ddns.py", line 14, in getMyIP
    ping = json.loads(requests.post(apiConfig["endpoint"] + '/ping/', data = json.dumps(apiConfig)).text)
  File "/usr/lib/python3.8/json/__init__.py", line 357, in loads
    return _default_decoder.decode(s)
  File "/usr/lib/python3.8/json/decoder.py", line 337, in decode
    obj, end = self.raw_decode(s, idx=_w(s, 0).end())
  File "/usr/lib/python3.8/json/decoder.py", line 355, in raw_decode
    raise JSONDecodeError("Expecting value", s, err.value) from None
json.decoder.JSONDecodeError: Expecting value: line 1 column 1 (char 0)

From k****u@the-server.T****h  Mon Jun 21 00:22:04 2021
Return-Path: <k****u@the-server.T****h>
X-Original-To: k****u
Delivered-To: k****u@the-server.T****h
Received: by the-server.T****h (Postfix, from userid 1000)
        id B2D57EC0AE; Mon, 21 Jun 2021 00:22:04 -0500 (CDT)
From: root@the-server.T****h (Cron Daemon)
To: k****u@the-server.T****h
Subject: Cron <k****u@the-server> /bin/python3 /External/porkbun-dynamic-dns-python/porkbun-ddns.py /External/porkbun-dynamic-dns-python/config.json z***k.family www -i ${myip}
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit
X-Cron-Env: <SHELL=/bin/sh>
X-Cron-Env: <HOME=/home/k****u>
X-Cron-Env: <PATH=/usr/bin:/bin>
X-Cron-Env: <LOGNAME=k****u>
Message-Id: <20210621052204.B2D57EC0AE@the-server.T****h>
Date: Mon, 21 Jun 2021 00:22:04 -0500 (CDT)

Deleting existing A Record
Creating record: www.z***k.family with answer of 67.***.***.217
Traceback (most recent call last):
  File "/External/porkbun-dynamic-dns-python/porkbun-ddns.py", line 50, in <module>
    print(createRecord()["status"])
  File "/External/porkbun-dynamic-dns-python/porkbun-ddns.py", line 28, in createRecord
    create = json.loads(requests.post(apiConfig["endpoint"] + '/dns/create/'+ rootDomain, data = json.dumps(createObj)).text)
  File "/usr/lib/python3.8/json/__init__.py", line 357, in loads
    return _default_decoder.decode(s)
  File "/usr/lib/python3.8/json/decoder.py", line 337, in decode
    obj, end = self.raw_decode(s, idx=_w(s, 0).end())
  File "/usr/lib/python3.8/json/decoder.py", line 355, in raw_decode
    raise JSONDecodeError("Expecting value", s, err.value) from None
json.decoder.JSONDecodeError: Expecting value: line 1 column 1 (char 0)

From k****u@the-server.T****h  Mon Jun 21 00:22:04 2021
Return-Path: <k****u@the-server.T****h>
X-Original-To: k****u
Delivered-To: k****u@the-server.T****h
Received: by the-server.T****h (Postfix, from userid 1000)
        id B8883EC0AE; Mon, 21 Jun 2021 00:22:04 -0500 (CDT)
From: root@the-server.T****h (Cron Daemon)
To: k****u@the-server.T****h
Subject: Cron <k****u@the-server> myip="$(dig +short myip.opendns.com @resolver1.opendns.com)" && /bin/python3 /External/porkbun-dynamic-dns-python/porkbun-ddns.py /External/porkbun-dynamic-dns-python/config.json z***k.io -i ${myip}
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit
X-Cron-Env: <SHELL=/bin/sh>
X-Cron-Env: <HOME=/home/k****u>
X-Cron-Env: <PATH=/usr/bin:/bin>
X-Cron-Env: <LOGNAME=k****u>
Message-Id: <20210621052204.B8883EC0AE@the-server.T****h>
Date: Mon, 21 Jun 2021 00:22:04 -0500 (CDT)

Deleting existing A Record
Creating record: z***k.io with answer of 67.***.***.217
SUCCESS
```
Thanks for making this, really appreciate it!

### LINKED ISSUE #3: DDNS for IPv6
Hi,
Would be possible to update the script for IPv6 support? As far I understand the script is supporting just IPv4.
Thanks

### PR DESCRIPTION
**Pull request by con-f-use**
Modernizes the Python DDNS client:

* Use argparse for CLI argument handling (better help text, validation and error messages)
* Accept config file from stdin
* Use modern Python naming conventions
* Use ipaddress for validation
* Default value for the base url of the API
* Hashbang and magic `__main__` for use as executable script
* Closes #2 
* Closes #3

### CODE REVIEW

**Comment by con-f-use** [comment]
@rossporkbun @eddie-porkbun please review. I tried to keep the CLI as much as possible - at some point, in a later PR, I'd like to add braking changes if that is okay with you:

Namely, a default config file location in `~/.config/porkbun/ddns.conf` that can be overwritten via environment variable and `-c, --config` option as well as the possibility to save domain and subdomain there and overwrite everything with their own command line options. Also: logging, verbosity option and maybe a `setup.py` and `flake.nix` for packaging and easy distribution. But only if you're interested.

**Comment by benneti** [comment]
If I try to use this with -i IPV6 adress, it set's the A record, I guess it should properly format the args when the ip is given.

**Comment by con-f-use** [comment]
@benneti are you sure, because there is: https://github.com/con-f-use/porkbun-dynamic-dns-python/blob/patch-1/porkbun-ddns.py#L52 in the code.

**Comment by benneti** [comment]
Thanks for the quick reply.
Yeah I saw this, too and was surprised that it nonetheless chose A.
I can double check later whats args contains if I use -i.

On 12 February 2022 11:02:58 CET, con-f-use ***@***.***> wrote:
***@***.*** are you use, because there is: https://github.com/con-f-use/porkbun-dynamic-dns-python/blob/patch-1/porkbun-ddns.py#L52 in the code.
>
>-- 
>Reply to this email directly or view it on GitHub:
>https://github.com/porkbundomains/porkbun-dynamic-dns-python/pull/6#issuecomment-1037101563
>You are receiving this because you were mentioned.
>
>Message ID: ***@***.***>
-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

**Comment by benneti** [comment]
ok sorry it works as intended now, I guess I mixed up the lines.
The actual problem is that if I set the AAAA record it also deletes the A record, but I would like to have both set at the same time.
Basically I think
```python
def delete_record(args):
    type_ = "A" if args.public_ip.version == 4 else "AAAA"
    for i in get_records(args)["records"]:
        if i["name"] == args.fqdn and i["type"] in [type_, "ALIAS", "CNAME"]:
            print("Deleting existing {}-Record: {}".format(i["type"], i))
            api(args, "/dns/delete/" + args.domain + "/" + i["id"])
```
would be better suited

File: porkbun-ddns.py [hunk 1]
```diff
@@ -1,53 +1,94 @@
...
+            args.domain,
+        )
+    return all_records
+
+
+def delete_record(args):
+    for i in get_records(args)["records"]:
+        if i["name"] == args.fqdn and i["type"] in ["A", "AAAA", "ALIAS", "CNAME"]:
+            print("Deleting existing {}-Record: {}".format(i["type"], i))
+            api(args, "/dns/delete/" + args.domain + "/" + i["id"])
```
Comment (con-f-use) [inline porkbun-ddns.py]:
```suggestion
def delete_record(args):
    type_ = "A" if args.public_ip.version == 4 else "AAAA"
    for i in get_records(args)["records"]:
        if i["name"] == args.fqdn and i["type"] in [type_, "ALIAS", "CNAME"]:
            print("Deleting existing {}-Record: {}".format(i["type"], i))
            api(args, "/dns/delete/" + args.domain + "/" + i["id"])
```

**Comment by con-f-use** [comment]
@benneti
Did you know, you can make code suggestions in comments:

image

You just need to be in "Files Changed" and then click on a staring line green plus icon and drag to the last line your change is about.

**Comment by benneti** [comment]
I did not, but I think I could also have done a review to achieve the same.
Thanks for putting it in the PR anyway!

**Comment by rossporkbun** [comment]
@con-f-use sorry for the delay in this response. Thank you so much for doing all this work, and I bow to your Python knowledge, it bests mine. 

I'm guessing this can't be made Python 2 compliant, which is a necessary requirement for the program. If you can, I'd be glad to consider the pull request. Otherwise, if you want to release the client on your own, I will happily link to your repo in our official documentation. That way, you can maintain it yourself rather than relying on me to approve your pull requests, as I am admittedly slow to take action.

**Comment by con-f-use** [comment]
I think it can be lightly tweaked to work with Python 2 AND 3. I'll see what I can do.

That being said, a word of advice: Python 2 has been end of life for close to ~two~ three years non and does not get any updates, security or otherwise. You should really switch any remaining infrastructure, because soon you'll face increasing problems with even installing it, since a lot of tools (pip, setuptools and the python package index) will or have already phased out support. I saw those problems daily in the company I work for.

**Comment by con-f-use** [comment]
It does work with python2.7 as long as [`py2-ipaddress`](https://pypi.org/project/py2-ipaddress/) and [`requests`](https://pypi.org/project/requests/2.27.1/#description) are installed in the right versions.

**Comment by mietzen** [comment]
@rossporkbun any Update on this? This would be highly beneficial for all users with IPv6 dual-stack lite connection which getting more common.

**Comment by dramborleg** [comment]
> @rossporkbun any Update on this? This would be highly beneficial for all users with IPv6 dual-stack lite connection which getting more common.

Given the low activity I'd be surprised to see a response in the near future, but just in case it happens to help you I figured I would mention that ddclient recently added porkbun support in https://github.com/ddclient/ddclient/issues/437. Not sure if it's in an official ddclient release yet.

**Comment by eddie-porkbun** [comment]
> @rossporkbun any Update on this? This would be highly beneficial for all users with IPv6 dual-stack lite connection which getting more common.

Ross actually has left Porkbun some time ago for new adventures. Ultimately this was meant to be an example of how to use our API for DDNS, provided as is, rather than an officially supported client. Using ddclient is probably a better way to go and I'll be sure to test it out myself soon.

**Comment by mietzen** [comment]
@eddie-porkbun 
I think it would be a good Idea to update the ReadMe, reference ddclient and archive this repo.
Since this repo is still one of the first hits when you Google ddns and porkbun.

**Pull request closed without merging**